components/brave_news/browser/combined_feed_parsing.cc

```diff
@@ -0,0 +1,146 @@
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_news/browser/combined_feed_parsing.h"
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/logging.h"
+#include "base/strings/strcat.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/time/time.h"
+#include "base/types/expected.h"
+#include "brave/components/brave_news/api/feed.h"
+#include "brave/components/brave_news/common/brave_news.mojom-forward.h"
+#include "brave/components/brave_news/common/brave_news.mojom-shared.h"
+#include "brave/components/brave_news/common/brave_news.mojom.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+#include "ui/base/l10n/time_format.h"
+#include "url/gurl.h"
+
+namespace brave_news {
+
+namespace {
+
+base::expected<mojom::FeedItemPtr, std::string> ParseFeedItem(
+    const base::Value& value) {
+  std::u16string error;
+  auto parsed_feed_item = api::feed::Item::FromValue(value, &error);
+  if (!parsed_feed_item) {
+    return base::unexpected(base::StrCat(
+        {"Failed to parse feed item. ", base::UTF16ToASCII(error)}));
+  }
+  api::feed::Item& feed_item = *parsed_feed_item;
+
+  auto url = GURL(feed_item.url);
+  if (url.is_empty() || !url.has_host()) {
+    return base::unexpected(base::StrCat(
+        {"Found feed item with an invalid url value. title=", feed_item.title,
+         ", url=", feed_item.url}));
+  }
+
+  if (!url.SchemeIsHTTPOrHTTPS()) {
+    return base::unexpected(
+        base::StrCat({"Item url was not HTTP or HTTPS: url=", url.spec()}));
+  }
+
+  if (feed_item.padded_img.empty()) {
+    return base::unexpected(base::StrCat(
+        {"Found feed item with missing image. url=", feed_item.url}));
+  }
+
+  if (feed_item.publisher_id.empty()) {
+    return base::unexpected(base::StrCat(
+        {"Found feed item with missing publisher id. url=", feed_item.url}));
+  }
+
+  if (feed_item.title.empty()) {
+    return base::unexpected(base::StrCat(
+        {"Found feed item with missing title. url=", feed_item.url}));
+  }
+
+  if (!feed_item.score.has_value()) {
+    // We just warn, as this is an optional field.
+    VLOG(1) << "Item was missing score: " << feed_item.url;
+  }
+
+  auto metadata = mojom::FeedItemMetadata::New();
+  metadata->category_name = feed_item.category;
+  metadata->title = feed_item.title;
+  metadata->description = feed_item.description;
+  metadata->publisher_id = feed_item.publisher_id;
+  metadata->publisher_name = feed_item.publisher_name;
+  metadata->image = mojom::Image::NewPaddedImageUrl(GURL(feed_item.padded_img));
+  metadata->url = std::move(url);
+
+  // Further weight according to history
+  metadata->score = feed_item.score.value_or(20.0);
+
+  // Extract time
+  if (!base::Time::FromUTCString(feed_item.publish_time.c_str(),
+                                 &metadata->publish_time)) {
+    VLOG(1) << "Bad time string for feed item: " << feed_item.publish_time;
```

Comment:
consider returning base::unexpected?

Reply:
I have discussed this with Pete, and his suggestion was that we shouldn't the parsing of the feed record because of this alone, but just log it.